reactioncommerce/api-plugin-products
# feat: add exact search option

Signed-off-by: Mohan Narayana <mohan.narayana@mailchimp.com>

Resolves #39 
Impact: **breaking|minor**
Type: **feature**

## Issue
Products metafield performs only fuzzy search. 

## Solution
Introduce the option to do exact match. The flag `isExactMatch` is added as an optional query parameter. It is by default set to false i.e., to perform a fuzzy search. It can be set to true to force exact search of metafields.

## Breaking changes
The default behavior to do fuzzy search will happen regardless of the new flag being used/set in the query. So no breaking change to the default behavior. When the flag is set to true, the query results can change.


## Testing
1. Create a few products and assign different metadata to them. Say {key=SKU, value= test1}, {key=SKU, value= test2} and so on.
2. Runt the query
```
query products ($shopId: ID!, 
  $metakey: String!, 
  $metavalue: String!) { 
    products (shopIds: [$shopId], 
      metafieldKey: $metakey, 
      metafieldValue: $metavalue) 
    { 
      totalCount
      nodes { 
        _id description title metafields {key value} 
        
      }
    }
}
```
with query variables
```
{
    "shopId": "<shopId>",
    "metakey": "SKU",
    "metavalue": "test"
}
```
It should show all the products with test* as metavalue. This is default behavior

3. Add the `isExactMatch` parameter with value `false`. This should still perform the default fuzzy search. Try using regex for metafields for further testing.
4. To test exact search set `isExactMatch` to true. Change the value of `metavalue` to see different search results. Using regex now should not return any results.
